src/result_processor.c

```diff
@@ -553,32 +553,41 @@ void SortAscMap_Dump(uint64_t tt, size_t n) {
 typedef struct {
   ResultProcessor base;
   uint32_t offset;
-  uint32_t limit;
-  uint32_t count;
+  uint32_t remaining;
 } RPPager;
 
-static int rppagerNext(ResultProcessor *base, SearchResult *r) {
+static int rppagerNext_Limit(ResultProcessor *base, SearchResult *r) {
   RPPager *self = (RPPager *)base;
-  int rc;
 
+  // If we've reached LIMIT:
+  if (!self->remaining) {
+    return RS_RESULT_EOF;
+  }
+
+  self->remaining--;
+  return base->upstream->Next(base->upstream, r);
+}
+
+static int rppagerNext_Skip(ResultProcessor *base, SearchResult *r) {
+  RPPager *self = (RPPager *)base;
+
+  // Currently a pager is never called more than offset+limit times.
+  // We limit the entire pipeline to offset+limit (upstream and downstream).
+  uint32_t limit = MIN(self->remaining, base->parent->resultLimit);
+  base->parent->resultLimit = self->offset + limit;
   // If we've not reached the offset
-  while (self->count < self->offset) {
+  while (self->offset) {
     int rc = base->upstream->Next(base->upstream, r);
     if (rc != RS_RESULT_OK) {
       return rc;
     }
-    self->count++;
+    base->parent->resultLimit--;
+    self->offset--;
     SearchResult_Clear(r);
   }
 
-  // If we've reached LIMIT:
-  if (self->count >= self->limit + self->offset) {
-    return RS_RESULT_EOF;
-  }
-
-  self->count++;
-  rc = base->upstream->Next(base->upstream, r);
-  return rc;
+  base->Next = rppagerNext_Limit; // switch to second phase
```

Comment:
Splitting the previous `rppagerNext` into two is in order to check the limit on the parent as well while skipping?